cpp/src/arrow/dataset/file_ipc_test.cc

```diff
@@ -151,9 +145,80 @@ TEST_F(TestIpcFileFormat, OpenFailureWithRelevantError) {
                                   result.status());
 }
 
-// TODO(bkietz) extend IpcFileFormat to support projection pushdown
-// TEST_F(TestIpcFileFormat, ScanRecordBatchReaderProjected)
-// TEST_F(TestIpcFileFormat, ScanRecordBatchReaderProjectedMissingCols)
+TEST_F(TestIpcFileFormat, ScanRecordBatchReaderProjected) {
+  schema_ = schema({field("f64", float64()), field("i64", int64()),
+                    field("f32", float32()), field("i32", int32())});
+
+  opts_ = ScanOptions::Make(schema_);
+  opts_->projector = RecordBatchProjector(SchemaFromColumnNames(schema_, {"f64"}));
+  opts_->filter = equal(field_ref("i32"), scalar(0));
+
+  // NB: projector is applied by the scanner; IpcFragment does not evaluate it so
+  // we will not drop "i32" even though it is not in the projector's schema
+  auto expected_schema = schema({field("f64", float64()), field("i32", int32())});
+
+  auto reader = GetRecordBatchReader();
+  auto source = GetFileSource(reader.get());
+  auto fragment = std::make_shared<IpcFragment>(*source, opts_);
+
+  int64_t row_count = 0;
+
+  for (auto maybe_batch : Batches(fragment.get())) {
+    ASSERT_OK_AND_ASSIGN(auto batch, std::move(maybe_batch));
+    row_count += batch->num_rows();
+    AssertSchemaEqual(*batch->schema(), *expected_schema,
+                      /*check_metadata=*/false);
+  }
+
+  ASSERT_EQ(row_count, kNumRows);
+}
+
+TEST_F(TestIpcFileFormat, ScanRecordBatchReaderProjectedMissingCols) {
+  schema_ =
+      schema({field("f64", float64()), field("i64", int64()), field("f32", float32())});
+  auto reader_without_i32 = GetRecordBatchReader();
+
+  schema_ =
+      schema({field("i64", int64()), field("f32", float32()), field("i32", int32())});
+  auto reader_without_f64 = GetRecordBatchReader();
+
+  schema_ = schema({field("f64", float64()), field("i64", int64()),
+                    field("f32", float32()), field("i32", int32())});
+  auto reader = GetRecordBatchReader();
+
+  opts_ = ScanOptions::Make(schema_);
+  opts_->projector = RecordBatchProjector(SchemaFromColumnNames(schema_, {"f64"}));
+  opts_->filter = equal(field_ref("i32"), scalar(0));
+
+  for (auto reader : {reader.get(), reader_without_i32.get(), reader_without_f64.get()}) {
+    auto source = GetFileSource(reader);
+    auto fragment = std::make_shared<IpcFragment>(*source, opts_);
+
+    int64_t row_count = 0;
+
+    for (auto maybe_batch : Batches(fragment.get())) {
+      ASSERT_OK_AND_ASSIGN(auto batch, std::move(maybe_batch));
+      row_count += batch->num_rows();
+      // NB: projector is applied by the scanner; ParquetFragment does not evaluate it.
+      // We will not drop "i32" even though it is not in the projector's schema.
+      //
+      // in the case where a file doesn't contain a referenced field, we won't
+      // materialize it (the filter/projector will populate it with nulls later)
+      std::shared_ptr<Schema> expected_schema;
+      if (reader == reader_without_i32.get()) {
```

Comment:
The schema should be set out of this loop.

Reply:
will do